File: packages/cli/src/ckpt/commands/push.py
"""ckpt push — push to remote and sync reasoning to API."""

from __future__ import annotations

import subprocess
import sys

import click

from ckpt.core import git
from ckpt.storage.local import LocalStore
from ckpt.sync.client import sync_reasoning


@click.command(
    context_settings={"ignore_unknown_options": True, "allow_extra_args": True},
)
@click.argument("remote", default="origin", required=False)
@click.argument("branch", default=None, required=False)
@click.pass_context
def push(ctx: click.Context, remote: str, branch: str | None):
    """Push to remote and sync reasoning records to the API."""
    # Build git push args
    git_args = ["push", remote]
    if branch:
        git_args.append(branch)
    git_args.extend(ctx.args)

    try:
        git.run(*git_args)
    except subprocess.CalledProcessError as e:
        sys.exit(e.returncode)

    # Sync unsynced reasoning records
    store = LocalStore()
    unsynced = store.get_unsynced()

    if not unsynced:
        click.echo("No reasoning records to sync.")
        store.close()
        return

    click.echo(f"Syncing {len(unsynced)} reasoning record(s)...")
    synced_ids = sync_reasoning(unsynced)

    if synced_ids:
        store.mark_synced(synced_ids)
        click.echo(f"Synced {len(synced_ids)} record(s) to API.")
    else:
        click.echo("Warning: Could not sync reasoning to API. Records will sync on next push.")

    store.close()

File: packages/cli/src/commands/push.ts
import { Command } from 'commander';
import { run, getRepoUrl } from '../core/git.js';
import { requireInit } from '../core/config.js';
import { LocalStore } from '../storage/local.js';
import { syncReasoning } from '../sync/client.js';

export const push = new Command('push')
  .description('Push to remote and sync reasoning')
  .argument('[remote]', 'remote name', 'origin')
  .argument('[branch]', 'branch name')
  .allowUnknownOption(true)
  .action(async (remote: string, branch: string | undefined, _opts: unknown, cmd: Command) => {
    requireInit();
    const gitArgs = ['push', remote];
    if (branch) gitArgs.push(branch);
    const extra = cmd.args?.filter((a: string) => a !== remote && a !== branch) ?? [];
    gitArgs.push(...extra);

    const result = run(gitArgs);
    if (result.exitCode !== 0) process.exit(result.exitCode);

    const store = new LocalStore();
    const unsynced = store.getUnsynced();

    if (unsynced.length === 0) {
      console.log('No reasoning records to sync.');
      store.close();
      return;
    }

    console.log(`Syncing ${unsynced.length} reasoning record(s)...`);
    const syncedIds = await syncReasoning(unsynced);

    if (syncedIds.length > 0) {
      store.markSynced(syncedIds);
      console.log(`Synced ${syncedIds.length} record(s).`);
    }

    store.close();
  });
